pypeclub/OpenPype
# General: Basic event system

## Brief description
Simple event system for triggering callbacks.

## Description
The event system should replace current avalon's system. Differences are few
- it should be possible to register methods as callbacks
- emitting of event won't just pass arguments to callbacks but will create `Event` object which is passed to callbac
    - `Event` object may contains data that can be used inside callback
- callback may have 1 or none arguments (event will be passed or not)
- callbacks can use star `*` is topics to be replaced with "any string"
    - callback registered to `"*"` will listen to all topics
    - it is possible to listen to partial topics `"workfile.*"` will be triggered for both `workfile.save` and `workfile.open`

## Changes
- implemented the event system inside `openpype/lib/events.`
    - public functions are `emit_event` and `register_event_callback`
- use new event system inside hosts and tools (replaces avalon's `on` and `emit`)

## Testing notes:
Almost all hosts are using even callbacks for initialization of catching workfile changes. They should still work as they did.

The easiest possible catch is using Loader tool and Workfiles tool because Workfiles tool is triggering `taskChanged` topic which is catched by Lodader.
1. Open any host where both tools are available
2. Open Loader - current asset should be selected in UI
3. Open Workfiles tool and change contest to other asset
4. Go back to Loader and click on "Go to current asset" button which should change asset selection in loader to current asset 
![image](https://user-images.githubusercontent.com/43494761/157022546-e7f00f04-4091-4f27-b117-527788127ae0.png)
- this will not work if the event is not triggered from workfiles tool and catched in loader
